src/encoder.rs

```diff
@@ -807,140 +808,580 @@ impl<W: Write> Drop for Writer<W> {
-    fn write(&mut self, mut buf: &[u8]) -> io::Result<usize> {
-        let written = buf.read(&mut self.buffer[self.index..])?;
-        self.index += written;
+    /// NOTE: this bypasses the internal buffer so the flush method should be called before this
+    ///       in the case there is some data left in the buffer when this is called, it will panic
+    fn write_header(&mut self) -> Result<()> {
+        if self.index != 0 {
+            // return Err(todo!())
+            panic!("Called when not flushed") // TODO: add error code
+        }
+        let wrt = self.writer.deref_mut();
 
-        if self.index >= self.buffer.len() {
-            self.writer
-                .as_mut()
-                .write_chunk(chunk::IDAT, &self.buffer)?;
-            self.index = 0;
+        if wrt.sep_def_img || wrt.info.frame_control.is_none() {
+            self.curr_chunk = chunk::IDAT;
+        } else {
+            if wrt.written == 0 {
+                self.curr_chunk = chunk::IDAT;
+            } else {
+                self.curr_chunk = chunk::fdAT;
+            }
+            if let Some(ref mut fctl) = wrt.info.frame_control {
+                fctl.encode(&mut wrt.w)?;
+                fctl.sequence_number += 1;
+            } else {
+                unreachable!()
+            }
         }
+        Ok(())
+    }
 
-        Ok(written)
+    /// Set the `FrameControl` for the following frame
+    ///
+    /// It will ignore the `sequence_number` of the parameter
+    /// as it is updated internally.
+    fn set_fctl(&mut self, f: FrameControl) {
+        if let Some(ref mut fctl) = self.writer.info.frame_control {
+            // ingnore the sequence number
+            *fctl = FrameControl {
+                sequence_number: fctl.sequence_number,
+                ..f
+            };
+        } else {
+            panic!("This function must be called on an animated PNG")
+        }
     }
 
-    fn flush(&mut self) -> io::Result<()> {
+    /// Flushes the current chunk
+    fn flush_inner(&mut self) -> io::Result<()> {
         if self.index > 0 {
-            self.writer
-                .as_mut()
-                .write_chunk(chunk::IDAT, &self.buffer[..self.index])?;
+            // flush the chunk and reset everything
+            write_chunk(
+                &mut self.writer.w,
+                self.curr_chunk,
+                &self.buffer[..self.index],
+            )?;
+            self.index = 0;
         }
-        self.index = 0;
         Ok(())
     }
 }
 
+impl<'a, W: Write> Write for ChunkWriter<'a, W> {
+    fn write(&mut self, mut data: &[u8]) -> io::Result<usize> {
+        if data.is_empty() {
+            return Ok(0);
+        }
+
+        // index == 0 means a chunk as been flushed out
+        if self.index == 0 {
+            let wrt = self.writer.deref_mut();
+            // ??? maybe use self.curr_chunk == chunk::fdAT ???
+            if !wrt.sep_def_img && wrt.info.frame_control.is_some() && wrt.written > 0 {
```

Comment:
Or add a field `header_written: bool`?

Reply:
If you think that's better/nicer I will do it

Reply:
Now that I read this again, I don't understand how it would be implemented that flag, as the `if` over there is not checking if the header is written or not but rather it's checking if the chunk whose header has already been written is an fdAT in order to add the sequence number.